### PR TITLE
PointOptimizerSummary option to avoid flattening

### PR DESCRIPTION
**Pull request by mpharrigan**
This isn't actually enough as the point optimizer will flatten it (again). But this *does* make it so I can write my own PointOptimizer replacement that can respect it

xref #2406

### CODE REVIEW

File: cirq/circuits/optimization_pass.py [hunk 1]
```diff
@@ -41,8 +41,11 @@ def __init__(self, clear_span: int, clear_qubits: Iterable['cirq.Qid'],
                 with each affected moment.
             new_operations: The operations to replace the cleared out
                 operations with.
+            preserve_moments: Whether to keep Moments intact instead of
+                flattening them
```
Comment (kevinsung) [inline cirq/circuits/optimization_pass.py]: Since, as you said, moments will still be flattened anyway, is this description a bit misleading?

Comment (mpharrigan) [reply cirq/circuits/optimization_pass.py]: updated description

File: cirq/circuits/optimization_pass.py [hunk 2]
```diff
@@ -41,8 +44,14 @@ def __init__(self, clear_span: int, clear_qubits: Iterable['cirq.Qid'],
                 with each affected moment.
             new_operations: The operations to replace the cleared out
                 operations with.
+            preserve_moments: Whether to keep Moments intact instead of
+                flattening them in this summary. Please be advised that
+                a PointOptimizer consuming this summary will flatten
+                operations no matter what, see gh-2406.
```
Comment (Strilanc) [inline cirq/circuits/optimization_pass.py]: Use a full link instead of the abbreviation.

File: cirq/circuits/optimization_pass.py [hunk 3]
```diff
@@ -41,8 +44,14 @@ def __init__(self, clear_span: int, clear_qubits: Iterable['cirq.Qid'],
                 with each affected moment.
             new_operations: The operations to replace the cleared out
                 operations with.
+            preserve_moments: Whether to keep Moments intact instead of
```
Comment (Strilanc) [inline cirq/circuits/optimization_pass.py]: To a user looking over this for the first time, this description is not very clear. I would rephrase into something like

```
If set, `cirq.Moment` instances within `new_operations` will be preserved exactly.
Normally the operations would be repacked to fit better into the target space, which
may move them between moments.
```

Comment (mpharrigan) [reply cirq/circuits/optimization_pass.py]: you should update the description on `flatten_op_tree`, as I have replicated it here :) Using your wording plus the additional warning per kevin's concerns

File: cirq/circuits/optimization_pass.py (hunk 2 above)
Comment (mpharrigan) [reply cirq/circuits/optimization_pass.py]: done
